esp/esdllib/esdl_xpath_extensions_libxml.cpp

```diff
@@ -472,6 +472,247 @@ static void strTokenizeFunction(xmlXPathParserContextPtr ctxt, int nargs)
+    xmlXPathReturnString(ctxt, xmlStrdup((const xmlChar *)encoded.str()));
+    xmlFree(toEncode);
+}
+
+static void unescapeXmlCharacters(xmlXPathParserContextPtr ctxt, int nargs)
+{
+    if (nargs != 1)
+    {
+        xmlXPathSetArityError(ctxt);
+        return;
+    }
+
+    xmlChar *toDecode = xmlXPathPopString(ctxt);
+    if (xmlXPathCheckError(ctxt)) //includes null check
+        return;
+
+    StringBuffer decoded;
+    decodeXML((const char*)toDecode, decoded);
+    xmlXPathReturnString(ctxt, xmlStrdup((const xmlChar *)decoded.str()));
+    xmlFree(toDecode);
+}
+
+static void compressString(xmlXPathParserContextPtr ctxt, int nargs)
+{
+    if (nargs != 1)
+    {
+        xmlXPathSetArityError(ctxt);
+        return;
+    }
+
+    xmlChar *toCompress = xmlXPathPopString(ctxt);
+    if (xmlXPathCheckError(ctxt))
+        return;
+
+    size32_t len = size32_t(strlen((const char *)toCompress)) + 1;
+    MemoryBuffer compressed;
+    compressToBuffer(compressed, len, (void *)toCompress);
+    StringBuffer encoded;
+    JBASE64_Encode((void*)compressed.bufferBase(), compressed.length(), encoded, false);
+    xmlXPathReturnString(ctxt, xmlStrdup((const xmlChar *)encoded.str()));
+
+    xmlFree(toCompress);
+}
+
+static void decompressString(xmlXPathParserContextPtr ctxt, int nargs)
+{
+    if (nargs != 1)
+    {
+        xmlXPathSetArityError(ctxt);
+        return;
+    }
+
+    xmlChar *toDecode = xmlXPathPopString(ctxt);
+    if (xmlXPathCheckError(ctxt))
+        return;
+    if (isEmptyString((const char*)toDecode))
+    {
+        xmlXPathSetError(ctxt, XPATH_EXPR_ERROR);
+        return;
+    }
+
+    MemoryBuffer toDecompress;
+    JBASE64_Decode((const char *)toDecode, toDecompress);
+    if (!toDecompress.length())
+    {
+        xmlXPathSetError(ctxt, XPATH_EXPR_ERROR);
+        return;
+    }
+
+    try
+    {
+        MemoryBuffer decompressed;
+        decompressToBuffer(decompressed, toDecompress);
+        xmlXPathReturnString(ctxt, xmlStrdup(decompressed.bytes()));
+    }
+    catch (IException* e)
+    {
+        e->Release();
+        xmlXPathSetError(ctxt, XPATH_EXPR_ERROR);
+    }
+    catch (...)
+    {
+        xmlXPathSetError(ctxt, XPATH_EXPR_ERROR);
+    }
+
+    xmlFree(toDecode);
+}
+
+static void toXmlString(xmlXPathParserContextPtr ctxt, int nargs)
+{
+    if (nargs != 1)
+    {
+        xmlXPathSetArityError(ctxt);
+        return;
+    }
+
+    xmlNodeSetPtr toSerialize = xmlXPathPopNodeSet(ctxt);
+    if (xmlXPathCheckError(ctxt))
+        return;
+
+    StringBuffer xml;
+    for (int i = 0; i < toSerialize->nodeNr; i++)
+    {
+        xmlNodePtr node = toSerialize->nodeTab[i];
+        if (!node)
+            continue;
+        xmlOutputBufferPtr xmlOut = xmlAllocOutputBuffer(nullptr);
+        xmlNodeDumpOutput(xmlOut, node->doc, node, 0, 1, nullptr);
+        xmlOutputBufferFlush(xmlOut);
+        xmlBufPtr buf = (xmlOut->conv != nullptr) ? xmlOut->conv : xmlOut->buffer;
```

Comment:
I'm cleanup not required for `xmlBufPtr buf` ?

Reply:
It should not be necessary. It is part of `xmlOut`, which is cleaned by `xmlOutputBufferClose`.